File: src/Systems/Swift/example3.swift
import Foundation

// Abstract Product
protocol Report {
    func generate()
}

// Concrete Product
class PDFReport: Report {
    func generate() {
        print("Generating PDF report")
    }
}

class HTMLReport: Report {
    func generate() {
        print("Generating HTML report")
    }
}

// Abstract Factory
protocol ReportFactory {
    func createReport() -> Report
}

// Concrete Factory
class PDFReportFactory: ReportFactory {
    func createReport() -> Report {
        return PDFReport()
    }
}

class HTMLReportFactory: ReportFactory {
    func createReport() -> Report {
        return HTMLReport()
    }
}

func useFactory(factory: ReportFactory) {
    let report = factory.createReport()
    report.generate()
}

// Usage
useFactory(factory: PDFReportFactory())
useFactory(factory: HTMLReportFactory())

File: src/Systems/Rust/example3.rs
trait Report {
    fn generate(&self);
}

struct PDFReport;

impl Report for PDFReport {
    fn generate(&self) {
        println!("Generating PDF report");
    }
}

struct HTMLReport;

impl Report for HTMLReport {
    fn generate(&self) {
        println!("Generating HTML report");
    }
}

trait ReportFactory {
    fn create_report(&self) -> Box<dyn Report>;
}

struct PDFReportFactory;

impl ReportFactory for PDFReportFactory {
    fn create_report(&self) -> Box<dyn Report> {
        Box::new(PDFReport)
    }
}

struct HTMLReportFactory;

impl ReportFactory for HTMLReportFactory {
    fn create_report(&self) -> Box<dyn Report> {
        Box::new(HTMLReport)
    }
}

fn use_factory(factory: &dyn ReportFactory) {
    let report = factory.create_report();
    report.generate();
}

fn main() {
    let pdf_factory = PDFReportFactory;
    let html_factory = HTMLReportFactory;
    use_factory(&pdf_factory);
    use_factory(&html_factory);
}
